Code/GraphMol/MolStandardize/Charge.cpp

```diff
@@ -332,49 +332,62 @@ ROMol *Uncharger::uncharge(const ROMol &mol) {
   for (const auto &match : q_matches) {
     q_matched += omol->getAtomWithIdx(match[0].second)->getFormalCharge();
   }
-  unsigned int n_matched = SubstructMatch(*omol, *(this->neg), n_matches);
-  unsigned int a_matched = SubstructMatch(*omol, *(this->neg_acid), a_matches);
+  SubstructMatch(*omol, *(this->neg), n_matches);
+  SubstructMatch(*omol, *(this->neg_acid), a_matches);
   std::vector<int> n_matches_mol;
-  for (const auto &n_match : n_matches) {
-    std::transform(n_match.begin(), n_match.end(),
-                   std::back_inserter(n_matches_mol),
-                   [](const std::pair<int, int> &p) { return p.second; });
-  }
+  std::transform(n_matches.begin(), n_matches.end(),
+                 std::back_inserter(n_matches_mol),
+                 [](const MatchVectType &mv) { return mv.front().second; });
+  unsigned int n_matched = n_matches_mol.size();
+  // a_matches may containe multiple degenerate matches
+  // (e.g., both [O-] in NO3-). We only want to keep the one with
+  // the lowest atom index
   std::vector<int> a_matches_mol;
+  std::sort(a_matches.begin(), a_matches.end(),
+            [](const MatchVectType &a, const MatchVectType &b) {
+              return (a.front().second < b.front().second);
+            });
+  std::vector<unsigned int> atomRanks(omol->getNumAtoms());
+  std::unordered_set<int> a_ranks;
+  Canon::rankMolAtoms(*omol, atomRanks, false, false, false);
```

Comment:
seems like we should only be paying the price for this step if `df_canonicalOrdering` is set.

Reply:
Here I use `atomRanks` to identify symmetry-equivalent groups to avoid taking into account the negative oxygens of `O=N([O-])[O-]` twice. In other words, I use `atomRanks` to filter symmetry-equivalent atoms from the `vector` of single-atom `MatchVectTyoe`s returned by `SubstructMatch`. The only alternative I see is to break the `heg_acid` filters into multiple non-recursive SMARTS and applying them sequentially. Happy to hear your thoughts if you have other ideas.